node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -269,6 +270,21 @@ struct WaitingCollationFetches {
 type ActiveCollationFetches =
 	FuturesUnordered<Pin<Box<dyn Future<Output = (Hash, PeerId)> + Send + 'static>>>;
 
+/// Core assignment type for a given relay parent.
+#[derive(Debug, Copy, Clone)]
+enum AssignmentType {
+	/// Current assignment refers to a common case of candidate being
+	/// backed in the child of relay parent. In particular, the core of
+	/// our parachain should be scheduled.
+	Current,
+	/// Lookahead assignment defines validators group for candidates being
+	/// backed in the grandchild of relay parent, therefore based on the child.
+	/// The core is expected to be occupied.
+	///
+	/// **Important note**: parathread cores very well may happen to be free.
```

Comment:
Tangentially related, but as you have studied this part of the code and thought about parathreads, what information do you think the runtime needs to provide about parathread cores in the future? Since we also want to experiment with other kinds of exotic scheduling (2 or 3 chains sharing 1 slot and taking turns, for instance), it would be nice to have a unified abstraction for scheduling information in the runtime API.

Reply:
For the current case when a parachain takes a core for the whole session duration it's fine to rely on scheduled-occupied-scheduled... behavior

With parathreads a core can happen e.g. to be free-scheduled-occupied, so we need to be aware in advance for which para id the core is going to be scheduled.

On the other hand, with async backing this change will be unnecessary, we could probably just drop it later